### ISSUE TITLE
Orleans silos crashing 'Process is terminating due to StackOverflowException.'

### ISSUE THREAD

**Issue by davebown** [dotnet/orleans#4939]
We have recently put an orleans service into production to manage the login session lifecycle of our application. This is running Orleans 2.0.4 and .net core 2.0 hosted in Kubernetes using Linux docker containers built using the microsoft/aspnetcore:2.0 image.

We are using Azure Storage for clustering and state based persistence..

Occasionally at first and with increased frequency we are seeing the silo host containers crash out due to a Stack Over Flow.

We believe this is somehow related to the use of 'reminders' because with a clean state for around the first 40 minutes there are no crashes but gradually over a number of days crashes and restarts happen more and more frequently until it is impossible to keep the cluster up and running in a health state.

While we are not ruling out our specific application code as the cause (accept this is the most likely) we can find no instances of recursive function calls, have experimented with removal and refactoring of several areas of the code and have so far been unsuccessful in resolving this.

This doesn't happen on our internal development and testing clusters but that isn't under load in the same way.

Are there any known issues / potential configurations with Orleans 2.x that are known to cause this? What information do I need to collect / how can i go about diagnosing this?

**Comment by sergeybykov** [comment]
In general, I can't remember seeing a SO exceptions in Orleans for years. There was a recent issue #4873, but that would manifest immediately at silo configuration time.

Can you get a stack trace or a memory dump when this happens? How many CPU cores do you give to each silo container?

**Comment by davebown** [comment]
I'll work on getting a memory dump but there isn't a stack trace in the container logs unfortunately.

At the moment i'm not requesting or limiting CPUs on the container, although we are with memory (currently 1gb request, 8gb limit) Is the lack of resource allocation on the CPUs likely to be a source of the problem? This is something that can easily be fixed but hasn't been necessary with other services with exception to databases like Elasticsearch and Casandra to help reduce latency.

**Comment by sergeybykov** [comment]
There is no hard rule here. However, for better performance and stability we recommend to run silos on 4+ core VMs/containers or physical servers, and 2 cores being the reasonable minimum. It would still run on a single-core VM/container, but once you push enough load to it, at some point there won't be enough parallelism in the system for it to stay responsive. That is still a far cry from SO exceptions that, like I said, we haven't seen in ages.

**Comment by sergeybykov** [comment]
@davebown Any update?

**Comment by davebown** [comment]
Unfortunately for now we've had to abandon Orleans for this use case. Pressure was on us to get a fix in place and we've ended up with a more traditional combination of api + redis backend for persistence. We don't have a lot of experience of analysing memory dumps etc and couldn't replicate outside production and the containers weren't set up to allow memory dumping in production. My suspicion given comments above and other conversations I've had is that Orleans just doesn't scale down to what we're trying to use it for i.e. persistence and domain logic in a relatively small sub-system of a micro-service, albeit one with potentially high throughput.  The Orleans model seemed conceptually to nicely fit the application processes we were trying to create but from an infrastructure perspective I think it's simply a mis-match.
